torch/csrc/autograd/python_function.cpp

```diff
@@ -381,90 +378,104 @@ static void _transplant_var(VariableImpl& var, const std::shared_ptr<Function>&
-        // current function to ensure the gradient is computed correctly.
-        // There are two cases now:
-        // 1. It has been modified in-place. If it didn't require_grad it's ok,
-        // but if it does, then it's a clear error.
-        // 2. It hasn't been modified. This means that it must have been
-        // returned unchanged, and we can simply return a new Variable
-        // referencing the same storage.
-        if (dirty_inputs.count(output) > 0) {
-          if (!input_var_.requires_grad()) {
-            Py_INCREF(input_var);
-            output_var = input_var;
-            _transplant_var(*input_var_.get(), cdata, i, is_volatile);
-          } else { // input_var_.requires_grad
-            throw std::runtime_error("a leaf Variable that requires grad has been used in an in-place operation.");
-          }
-        } else {
-          // An input has been returned, but it wasn't modified. It's better
-          // not to move the Variable, because there are some legitimate cases
-          // where making it non-leaf would break stuff (e.g. broadcast). Also,
-          // returning the input Variable is not a good option either,
-          // because if someone registers hooks on it, they will fire with grads
-          // from all usages, not only from usages of this output. This is why
-          // we'll return a copy and join their version counters. This has
-          // a side-effect of making in-place ops on any of these Variables an
-          // immediate error, but it would be raised anyway once someone
-          // calls backward.
-          if (is_volatile) {
-            output_var = (THPVariable*)THPVariable_NewVolatile(output);
-          } else {
-            output_var = (THPVariable*)THPVariable_NewWithFunction(output, cdata);
-          }
-          if (!output_var) throw python_error();
-          output_var->cdata.version_counter() = input_var->cdata.version_counter();
+      if (dirty_inputs.count(output) > 0) {
+        if (is_leaf && var.requires_grad()) {
+          throw std::runtime_error("a leaf Variable that requires grad has been used in an in-place operation.");
         }
+        // If the input was modified, transplant the grad_fn in the graph:
+        // grad_fn <- variable <- self  ==>  grad_fn <- self <- variable
+        _transplant_var(var, cdata, i, is_volatile);
+      } else if (is_leaf) {
+        // An input has been returned, but it wasn't modified. It's better
+        // not to move the Variable, because there are some legitimate cases
+        // where making it non-leaf would break stuff (e.g. broadcast). Also,
+        // returning the input Variable is not a good option either,
+        // because if someone registers hooks on it, they will fire with grads
+        // from all usages, not only from usages of this output. This is why
+        // we return a view on the input Variable.
+        var = new_variable(output);
```

Comment:
It seems to me that if we get an input Variable that wasn't a leaf **and** wasn't modified, then it won't enter any of those cases. Can you add a comment to show that we thought about this case as well (there was one for it in the old code)?